### PR TITLE
[release/2.x] Backports support for election while reconfiguration 

### PR DESCRIPTION
**Pull request by jumaffre**
Backports the following commits to release/2.x:
- https://github.com/microsoft/CCF/pull/4041
- https://github.com/microsoft/CCF/pull/4018
- https://github.com/microsoft/CCF/pull/4047

Note that CheckQuorum Raft extension is not included in 2.x releases.

### CODE REVIEW

File: tests/partitions_test.py [hunk 1]
```diff
@@ -261,6 +262,92 @@ def test_new_joiner_helps_liveness(network, args):
...
+    LOG.info("Isolate original backups and issue reconfiguration of another quorum")
+    with network.partitioner.partition(backups):
+        LOG.info("Trust all new nodes in one single proposal")
+        # Note: Commit is stuck since a majority of backups in initial configuration
+        # are isolated
+        network.consortium.trust_nodes(
+            primary,
+            [n.node_id for n in new_nodes],
+            valid_from=datetime.utcnow(),
+            wait_for_commit=False,
```
Comment (jumaffre) [inline tests/partitions_test.py:298]:
```suggestion
            wait_for_global_commit=False,
```

**Review by achamayou — APPROVED** [review]

**Comment by jumaffre** [comment]
This PR also contains [a fix](https://github.com/microsoft/CCF/pull/4076/commits/b45c06cf4a5008db2057a079bac276cfae85f73f) for the 1.x to 2.x code upgrade that started to fail on the `release/2.x` branch since https://github.com/microsoft/CCF/pull/4008.

**Pull request merged**